rfcs/convergence/react-focus-management.md

```diff
@@ -0,0 +1,169 @@
+
+Patterns in [ability-helpers](https://github.com/microsoft/ability-helpers) are **declarative**, and therefore decoupled more easily with framework and application code. Groups of focusables are managed by declaring `data-*` attributes which the lower level utilties use to define focusable behaviour.
+
+Nesting focusables is a problem that ability-helpers explicitly tries to solve, bringing focus management to a lower level than the rendering framework tries to accomplish this consistently independent of rendering cycles.
+
+### Initial API
+
+#### FocusManagementProvider
+
+Context provider that hosts the [ability-helpers](https://github.com/microsoft/ability-helpers) instance and its API objects.
+
+The below table defines what is included in the context value
+
+> Note that this table will increase as more ability-helpers features are included in Fluent
+
+| Name      | Value        | Description                           |
+| --------- | ------------ | ------------------------------------- |
+| focusable | FocusableAPI | API for traversing focusable elements |
+
+The below table defines the props used by the provider component. All context values can be used as props in the provider component but are optional.
+
+| Name                   | Value              | Description                                                                                                           |
+| ---------------------- | ------------------ | --------------------------------------------------------------------------------------------------------------------- |
+| dir                    | 'ltr' \| 'rtl'     | Direction indicator, some keyboard behaviour will need to change based on this value                                  |
+| Deloser                | DeloserAPI         | Tools to help automatically restore focus when it gets lost without manually calling `.focus()` from application code |
+| internal\_\_ahInstance | AbilityHelpersCore | The ability helpers instance, not expected to be be used directly                                                     |
+
+#### Hooks
+
+This spec proposes an API based on React hooks pattern. [ability-helpers](https://github.com/microsoft/ability-helpers) works mainly on adding `data-*` attributes to DOM elements to apply configurable behaviours to the 'groups' of focusables.
+
+However some internal behaviours such as `modalizer` or `deloser` require DOM elements during the configuration phase, which will be most easily configured in React using `refs`.
+
+Therefore it must be noted that **although some hooks might not require ref interaction and can simply return attributes to be spread, in an effort to maintain consistency refs are used, which fits the Fluent component design pattern which enforces ref forwarding**
+
+```typescript
+// Basic up/down keyboard navigation
+const ref = useKeyboardNavigationGroup();
```

Comment:
Do I correctly get that a hook will apply `data-*` attributes directly on DOM and will overwrite user's input?

```tsx
function useKeyboardNavigationGroup() {
  const ref = React.useRef();

  React.useRef(() => {
    const el = ref.current;

    el.dataset.someDataAttr = "someValue";
    // etc.
  });

  return ref;
}
```

Reply:
Yes, the motivation behind it is interesting....


AH has 'basic' properties which can be applied to the DOM as data-* attributes (there is a specific prefix for AH attributes just can't recall what is is exactly). In an idea scenario we shouldn't need refs at all...

but there APIs such as focus delosing which uses `onFocusLost` callback which is a non-serializable extended prop that needs a DOM node to apply correctly. Hence the current decision to use `refs` but I am open to change after some initial experiences

Reply:
No more refs for now, we'll deal with extended props when it gets to that point